gpflow/optimizers/natgrad.py

```diff
@@ -120,6 +120,107 @@ def naturals_to_xi(nat1, nat2):
+    xi_transform: XiTransform,
+):
+    """
+    This function does the backward step on the q_mu and q_sqrt parameters,
+    given the gradients of the loss function with respect to their unconstrained
+    variables. I.e., it expects the arguments to come from
+
+        with tf.GradientTape() as tape:
+            loss = loss_function()
+        q_mu_grad, q_mu_sqrt = tape.gradient(loss, [q_mu, q_sqrt])
+
+    (Note that tape.gradient() returns the gradients in *unconstrained* space!)
+
+    Implements equation [10] from
+
+    @inproceedings{salimbeni18,
+        title={Natural Gradients in Practice: Non-Conjugate Variational Inference in Gaussian Process Models},
+        author={Salimbeni, Hugh and Eleftheriadis, Stefanos and Hensman, James},
+        booktitle={AISTATS},
+        year={2018}
+
+    In addition, for convenience with the rest of GPflow, this code computes ∂L/∂η using
+    the chain rule (the following assumes a numerator layout where the gradient is a row
+    vector; note that TensorFlow actually returns a column vector), where L is the loss:
+
+    ∂L/∂η = (∂L / ∂[q_mu, q_sqrt])(∂[q_mu, q_sqrt] / ∂η)
+
+    In total there are three derivative calculations:
+    natgrad of L w.r.t ξ  = (∂ξ / ∂θ) [(∂L / ∂[q_mu, q_sqrt]) (∂[q_mu, q_sqrt] / ∂η)]ᵀ
+
+    Note that if ξ = θ (i.e. [q_mu, q_sqrt]) some of these calculations are the identity.
+    In the code η = eta, ξ = xi, θ = nat.
+
+    :param q_mu_grad: gradient of loss w.r.t. q_mu (in unconstrained space)
+    :param q_sqrt_grad: gradient of loss w.r.t. q_sqrt (in unconstrained space)
+    :param q_mu: parameter for the mean of q(u) with shape [M, L]
+    :param q_sqrt: parameter for the square root of the covariance of q(u)
+        with shape [L, M, M] (the diagonal parametrization, q_diag=True, is NOT supported)
+    :param xi_transform: the ξ transform to use (self.xi_transform if not specified)
+    """
+    _assert_shapes(q_mu, q_sqrt)
+
+    # 1) the ordinary gpflow gradient
+    dL_dmean = _to_constrained(q_mu_grad, q_mu.transform)
+    dL_dvarsqrt = _to_constrained(q_sqrt_grad, q_sqrt.transform)
+
+    with tf.GradientTape(persistent=True, watch_accessed_variables=False) as tape:
+        tape.watch([q_mu.unconstrained_variable, q_sqrt.unconstrained_variable])
+
+        # the three parameterizations as functions of [q_mu, q_sqrt]
+        eta1, eta2 = meanvarsqrt_to_expectation(q_mu, q_sqrt)
+        # we need these to calculate the relevant gradients
+        meanvarsqrt = expectation_to_meanvarsqrt(eta1, eta2)
+
+        if not isinstance(xi_transform, XiNat):
+            nat1, nat2 = meanvarsqrt_to_natural(q_mu, q_sqrt)
+            xi1_nat, xi2_nat = xi_transform.naturals_to_xi(nat1, nat2)
+            dummy_tensors = tf.ones_like(xi1_nat), tf.ones_like(xi2_nat)
+            with tf.GradientTape(watch_accessed_variables=False) as forward_tape:
+                forward_tape.watch(dummy_tensors)
+                dummy_gradients = tape.gradient(
```

Comment:
I'm not sure why it's showing that L190-L195 aren't covered by tests... I have checked and these lines do get reached by the current suite of tests? @st--